Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+to enable or disable keywords or levels. The currently enabled keywords and levels are encoded as
+bit vectors so that querying whether an event is enabled is as fast as loading three globals, a pointer
+dereference, and two bit tests, e.g. (hypothetically):
+
+```c++
+uint32_t enabledLevels;
+uint32_t enabledKeywords;
+
+bool GCEventStatus::IsEventEnabled(const GCEvent& event)
+{
+    return (enabledLevels & event.level) && (enabledKeywords & event.keyword);
+}
+```
+
+## Firing Events
+
+In order to fire an event, the GC will need to communicate with the EE in some way. The EE is ultimately
+responsible for routing the event to any appropriate subsystems (ETW, LTTNG, EventPipe) and the GC has
+no knowledge of what it is going to do with events that we give it.
+
+Given that we have validated that an event is enabled (through `GCEventStatus::IsEventEnabled`), we can
+fire the event using this new method on `IGCToCLR`:
+
+```c++
+void IGCToCLR::FireKnownEvent(
+    /* IN */ void* eventHandle,
+    /* IN */ void** eventData,
+    /* IN */ uint32_t eventDataCount,
+    /* IN */ const GUID* pActivityId = nullptr,
+    /* IN */ const GUID* pRelatedActivityId = nullptr
+);
+```
+
+The four arguments of this function are:
+1. `eventHandle` - An opaque handle to the EE's representation of the event being fired.
+2. `eventData` - An array of pointers that contains the payload of the event being fired.
+Each `GCEventData` is a tuple of a data pointer and size, so it can be used to describe arbitrary data.
+3. `eventDataCount` - The size of the `eventData` array.
+4. `pActivityId` and `pRelatedActivityId` - If provided, passes the activity ID and related activity ID
+verbatim to the underlying event implementation.
+
+The EE will use this information to determine 1) if this event truly is enabled and, 2) if it is,
+forwarding it to the apporopriate underlying event implementation, be it EventPipe, ETW, or LTTNG.
+The passed-in eventHandle will be a handle to an object from which the EE can get information about
+the version of the event and other useful metadata.
+
+## Getting Informed of Changes to Event State
+
+There are three mechanisms by which CoreCLR is able to log events: EventPipe, ETW, and LTTNG. When it
+comes to changing the state of events, EventPipe and ETW both allow users to attach callbacks that are
+invoked whenever events are enabled or disabled. For these two mechanisms, it is sufficient to use
+this callback mechanism to call `IGCHeap::{Enable/Disable}Events` from within such a callback in order
+to inform the GC of changes to tracing state.
+
+LTTNG does not have such a mechanism. In order to observe changes in the eventing state, LTTNG must be
+polled periodically. For this purpose, when tracing using LTTNG, a new thread will be spawned by the EE
+whose responsibility is to periodically poll LTTNG and invoke the GC whenever it detects changes in the
```

Comment:
This problem exists today. It is independent of standalone GC - it does not seem to be GC specific. I assume that this will be done for LTTNG accross the board.

Reply:
Today the EE relies entirely on LTTNG to answer the question of whether or not a specific event is enabled (through `tracepoint_enabled`). The EE does not need to be informed of changes to event state because LTTNG handles all of that. A standalone GC will not be invoking LTTNG directly and thus can't ask LTTNG whether or not a particular tracepoint is enabled, so I think this is a new problem.

Reply:
> Today the EE relies entirely on LTTNG

That's correct and it does not work well. #14327 has a long discussion about this. 

In particular, https://github.com/dotnet/coreclr/issues/14327#issuecomment-338314624 describes polling scheme very much like one that you have described.

Reply:
Ah, I see, thanks! I wasn't aware of that context. In that case it looks like we can piggyback on that same solution.

Reply:
Yup, it's reasonable to do this for all events on the EE side.  GC will want to shuttle this data across the interface as well to avoid calling into the EE to check event status.